R/metrics-ordinal.R

```diff
@@ -0,0 +1,160 @@
+#' @description Function assesses whether the inputs correspond to the
+#' requirements for scoring ordinal forecasts.
+#' @param observed Input to be checked. Should be an ordered factor of length n
+#'   with N levels holding the observed values. n is the number of observations
+#'   and N is the number of possible outcomes the observed values can assume.
+#' @param predicted Input to be checked. Should be nxN matrix of predicted
+#'   probabilities, n (number of rows) being the number of data points and N
+#'   (number of columns) the number of possible outcomes the observed values
+#'   can assume.
+#'   If `observed` is just a single number, then predicted can just be a
+#'   vector of size N.
+#'   Values represent the probability that the corresponding value
+#'   in `observed` will be equal to factor level referenced in `predicted_label`.
+#' @param predicted_label Ordered factor of length N with N levels, where N is
+#'   the number of possible outcomes the observed values can assume.
+#' @importFrom checkmate assert_factor assert_numeric assert_set_equal
+#' @inherit document_assert_functions return
+#' @keywords internal_input_check
+assert_input_ordinal <- function(observed, predicted, predicted_label) {
+  # observed
+  assert_factor(observed, min.len = 1, min.levels = 2, ordered = TRUE)
+  levels <- levels(observed)
+  n <- length(observed)
+  N <- length(levels)
+
+  # predicted label
+  assert_factor(
+    predicted_label,
+    len = N,
+    any.missing = FALSE, empty.levels.ok = FALSE, ordered = TRUE
+  )
+  if (!identical(levels(predicted_label), levels(observed))) {
+    cli_abort(
+      "Levels of `predicted_label` and `observed` must be identical
+      and in the same order. Found levels {.val {levels(predicted_label)}}
+      and {.val {levels(observed)}}."
+    )
+  }
+
+  # predicted
+  assert_numeric(predicted, min.len = 1, lower = 0, upper = 1)
+  if (n == 1) {
+    assert(
+      # allow one of two options
+      check_vector(predicted, len = N),
+      check_matrix(predicted, nrows = n, ncols = N)
+    )
+    summed_predictions <- .rowSums(predicted, m = 1, n = N, na.rm = TRUE)
+  } else {
+    assert_matrix(predicted, nrows = n)
+    summed_predictions <- round(rowSums(predicted, na.rm = TRUE), 10) # avoid numeric errors
+  }
+  if (!all(summed_predictions == 1)) {
+    #nolint start: keyword_quote_linter object_usage_linter
+    row_indices <- as.character(which(summed_predictions != 1))
+    cli_abort(
+      c(
+        `!` = "Probabilities belonging to a single forecast must sum to one",
+        `i` = "Found issues in row{?s} {row_indices} of {.var predicted}"
+      )
+    )
+    #nolint end
+  }
+  return(invisible(NULL))
+}
+
+
+#' Log score for ordinal outcomes
+#'
+#' @description
+#' **Log score for ordinal outcomes**
+#'
+#' The Log Score is the negative logarithm of the probability
+#' assigned to the observed value. It is a proper scoring rule. Small values
```

Comment:
is this what you mean or do you mean bigger i.e closer to zero because its negative?